// png-img/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PngImg = void 0;
//@ts-expect-error because it is a binary file
const png_img_1 = require("../build/Release/png_img");
const utils_1 = __importDefault(require("./utils"));
class PngImg {
    /**
     * Create PngImg object from passed buffer with image
     */
    constructor(rawImg) {
        this.img_ = new png_img_1.PngImg(rawImg);
    }
    /**
     * Get image size as an object.
     */
    size() {
        return {
            width: this.img_.width,
            height: this.img_.height
        };
    }
    /**
     * Get pixel color and alpha.
     * @param x x coordinate (left to right)
     * @param y y coordinate (top to bottom)
     */
    get(x, y) {
        return this.img_.get(x, y);
    }
    /**
     * Set pixel color
     * Same as fill(x, y, 1, 1, color)
     * (shorthand)
     * @param x x coordinate (left to right)
     * @param y y coordinate (top to bottom)
     * @param color color as rgb object or as a '#XXXXXX' string
     */
    set(x, y, color) {
        return this.fill(x, y, 1, 1, color);
    }
    /**
     * Fill region with passed color. Modifies current image.
     * @param offsetX offset from left side of the image
     * @param offsetY offset from top side of the image
     * @param x x coordinate (left to right)
     * @param y y coordinate (top to bottom)
     * @param color color as rgb object or as a '#XXXXXX' string
     */
    fill(offsetX, offsetY, width, height, color) {
        if (typeof color === 'string') {
            const objColor = utils_1.default.stringToRGBA(color);
            if (!objColor) {
                throw new Error('Bad color ' + color);
            }
            return this.fill(offsetX, offsetY, width, height, objColor);
        }
        color = {
            r: color.r || 0,
            g: color.g || 0,
            b: color.b || 0,
            a: color.a === undefined ? 255 : color.a
        };
        this.img_.fill(offsetX, offsetY, width, height, color);
        return this;
    }
    /**
     * Crop image. Modifies current image.
     * Throws if new image is not inside the current image.
     * @param offsetX offset from left side of the image
     * @param offsetY offset from top side of the image
     */
    crop(offsetX, offsetY, width, height) {
        this.img_.crop(offsetX, offsetY, width, height);
        return this;
    }
    /**
     * Sets new image size. Modifies current image.
     * If new size is less or equal than current size, then crop will be performed.
     */
    setSize(width, height) {
        const size = this.size();
        if (width <= size.width && height <= size.height) {
            return this.crop(0, 0, width, height);
        }
        this.img_.setSize(width, height);
        return this;
    }
    /**
     * Inserts image into specified place.
     * @param img image to insert
     * @param offsetX offset from left side of the image
     * @param offsetY offset from top side of the image
     */
    insert(img, offsetX, offsetY) {
        if (!(img instanceof PngImg)) {
            throw new Error('Not a PngImg object');
        }
        const imgSize = img.size();
        const mySize = this.size();
        if (offsetX + imgSize.width > mySize.width || offsetY + imgSize.height > mySize.height) {
            throw new Error('Out of the bounds');
        }
        this.img_.insert(img.img_, offsetX, offsetY);
        return this;
    }
    /**
     * Rotates image 90 degrees clockwise
     */
    rotateRight() {
        this.img_.rotateRight();
        return this;
    }
    /**
     * Rotates image 90 degrees counterclockwise
     */
    rotateLeft() {
        this.img_.rotateLeft();
        return this;
    }
    /**
     * Save image to file
     */
    async save(file) {
        return new Promise((resolve, reject) => {
            this.img_.write(file, (err) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
}
exports.PngImg = PngImg;
